pkg/source/kafka/kafka_source.go

```diff
@@ -0,0 +1,361 @@
+	// offset, or when calling ConsumePartition to start consuming from the
+	// oldest offset that is still available on the broker.
+	saramaConfig.Consumer.Offsets.Initial = cfg.OffsetsInitial
+
+	// Kafka rebalance strategy, defaulted to "range"
+	switch cfg.Assignor {
+	case "sticky":
+		saramaConfig.Consumer.Group.Rebalance.GroupStrategies = []sarama.BalanceStrategy{
+			sarama.BalanceStrategySticky,
+		}
+	case "roundrobin":
+		saramaConfig.Consumer.Group.Rebalance.GroupStrategies = []sarama.BalanceStrategy{
+			sarama.BalanceStrategyRoundRobin,
+		}
+	default:
+		saramaConfig.Consumer.Group.Rebalance.GroupStrategies = []sarama.BalanceStrategy{
+			sarama.BalanceStrategyRange,
+		}
+	}
+
+	// validate SASL if enabled
+	if cfg.EnableSASL {
+		saramaConfig.Net.SASL.Enable = true
+		saramaConfig.Net.SASL.User = cfg.SASLUsername
+		saramaConfig.Net.SASL.Password = cfg.SASLPassword
+		saramaConfig.Net.SASL.Handshake = true
+		if cfg.SASLAlgorithm == "sha512" {
+			saramaConfig.Net.SASL.SCRAMClientGeneratorFunc = func() sarama.SCRAMClient { return &xdgSCRAMClient{HashGeneratorFcn: SHA512} }
+			saramaConfig.Net.SASL.Mechanism = sarama.SASLTypeSCRAMSHA512
+		} else if cfg.SASLAlgorithm == "sha256" {
+			saramaConfig.Net.SASL.SCRAMClientGeneratorFunc = func() sarama.SCRAMClient { return &xdgSCRAMClient{HashGeneratorFcn: SHA256} }
+			saramaConfig.Net.SASL.Mechanism = sarama.SASLTypeSCRAMSHA256
+		} else if cfg.SASLAlgorithm == "plaintext" {
+			saramaConfig.Net.SASL.Mechanism = sarama.SASLTypePlaintext
+		} else {
+			return nil, fmt.Errorf("invalid SHA algorithm \"%s\": can be either \"sha256\" or \"sha512\"", cfg.SASLAlgorithm)
+		}
+	}
+
+	// validate TLS if required
+	tlsConfig, err := common.CreateTLSConfiguration(cfg.TLSCert, cfg.TLSKey, cfg.TLSCa, cfg.SkipVerifyTLS)
+	if err != nil {
+		return nil, err
+	}
+	if tlsConfig != nil {
+		saramaConfig.Net.TLS.Config = tlsConfig
+		saramaConfig.Net.TLS.Enable = true
+	}
+
+	client, err := sarama.NewConsumerGroup(strings.Split(cfg.Brokers, ","), fmt.Sprintf(`%s-%s`, cfg.ConsumerName, cfg.TopicName), saramaConfig)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to create Kafka client")
+	}
+
+	return newKafkaSourceWithInterfaces(client, &kafkaSource{
+		brokers:          cfg.Brokers,
+		topic:            cfg.TopicName,
+		consumerName:     cfg.ConsumerName,
+		log:              logger,
+		concurrentWrites: cfg.ConcurrentWrites,
+	})
+}
+
+// newKafkaSourceWithInterfaces creates a new source for reading messages from Apache Kafka, allowing the user to provide a mocked client.
+func newKafkaSourceWithInterfaces(client sarama.ConsumerGroup, s *kafkaSource) (*kafkaSource, error) {
+	s.client = client
+	return s, nil
+}
+
+// GetID returns the identifier for this target
+func (ks *kafkaSource) GetID() string {
+	return fmt.Sprintf("brokers:%s:topic:%s", ks.brokers, ks.topic)
+}
+
+func getKafkaVersion(targetVersion string) (sarama.KafkaVersion, error) {
```

Comment:
This is already defined in the target, we can factor it out into `pkg/common/helpers.go`